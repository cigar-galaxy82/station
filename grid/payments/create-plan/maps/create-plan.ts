/* eslint-disable jest/no-export */

import { describe, expect } from '@jest/globals';
import { SuperfaceTest } from '@superfaceai/testing';

import { createProduct } from '../../create-product/maps/create-product';

export async function createPlan(provider: string): Promise<string> {
  const productId = await createProduct(provider);

  const superfacePublishingProfiles = new SuperfaceTest({
    profile: 'payments/create-plan',
    provider,
  });
  const result = await superfacePublishingProfiles.run(
    {
      useCase: 'CreatePlan',
      input: {
        productId,
        name: 'Temporary Plan',
        interval: 'month',
        price: 1234,
        currency: 'USD',
      },
      testName: 'payments/create-plan/temporary',
    },
    {
      hideInput: ['productId'],
    }
  );

  return (result.unwrap() as { id: string }).id;
}

export function createPlanTest(providerName: string): void {
  describe(`payments/create-plan/${providerName}`, () => {
    let superface: SuperfaceTest;

    beforeEach(() => {
      superface = new SuperfaceTest({
        profile: 'payments/create-plan',
        provider: providerName,
        testInstance: expect,
      });
    });

    describe('CreatePlan', () => {
      describe('when all inputs are correct', () => {
        it('creates a plan for a product', async () => {
          const productId = await createProduct(providerName);
          const result = await superface.run(
            {
              useCase: 'CreatePlan',
              input: {
                productId,
                name: 'A Man, A Plan, A Canal, Panama',
                interval: 'month',
                price: 1234,
                currency: 'USD',
              },
            },
            {
              hideInput: ['productId'],
            }
          );
          expect(() => result.unwrap()).not.toThrow();
          expect(result.unwrap()).toMatchSnapshot();
        });
      });
    });
  });
}
