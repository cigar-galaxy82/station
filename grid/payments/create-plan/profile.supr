name = "payments/create-plan"
version = "1.0.0"

"""
Create plan usecase
"""

usecase CreatePlan unsafe {
  input {
    """
    Product ID
    The ID of the product to create a subscription plan for.
    (To be created with `payments/create-product`)
    """
    productId! string!

    """
    Name
    Name of the subscription plan
    """
    name! string!

    """
    Interval
    Frequency of the interval at which the customer is charged.

    Possible values:
    - `day`
    - `week`
    - `month`
    - `year`
    """
    interval! enum {
      day, week, month, year
    }!

    """
    Price
    Amount charged each `interval` cycle.
    """
    price! number!

    """
    Currency code
    Three-letter currency code in which `price` is specified, in ISO 4217 format.
    """
    currency! string!
 }

  result {
    """
    Plan ID
    ID of plan.
    """
    id! string!
    
    """
    State
    State of the plan.

    Possible values:
     - `created` - The plan was just created. It may not be possible to accept new subscriptions yet.
     - `active` - The plan is active and accepting new subscriptions.
     - `inactive` - The plan has been deactivated, and cannot accept new subscriptions.
    """
    state! enum {
      created, active, inactive
    }
  }

  error {
    """
    Title
    A short, human-readable summary of the problem type.
    """
    title! string

    """
    Detail
    A human-readable explanation specific to this occurrence of the problem.
    """
    detail string
  }

  example Successful {
    input {
      productId = "PROD-1192810198"
      name = "Monthly subscription to Superface milk delivery"
      interval = "month"
      price = 25
      currency = "USD"
    }

    result {
      id = "PLAN-89101829211"
      state = "active"
    }
  }
}