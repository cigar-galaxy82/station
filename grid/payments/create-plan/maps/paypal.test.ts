import { createPlanTest } from './create-plan';

createPlanTest('paypal');
