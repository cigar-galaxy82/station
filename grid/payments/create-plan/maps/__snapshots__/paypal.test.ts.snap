// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`payments/create-plan/paypal CreatePlan when all inputs are correct creates a plan for a product 1`] = `
Object {
  "id": "P-69E87546G1239230MMNL7U6A",
  "state": "active",
}
`;
